getThumbnail does not handle all encodings in TileOutputMimeTypes

The following test and note it fails for some encodings listed as supported

Note that this is causing failures for `django-large-image`: https://github.com/girder/django-large-image/runs/8082224710?check_suite_focus=true

```py
import pytest
import large_image
from large_image.constants import TileOutputMimeTypes


@pytest.fixture
def src():
    return large_image.open('/Users/bane/Desktop/rasters/converted.tif')


@pytest.mark.parametrize('encoding', TileOutputMimeTypes.keys())
def test_thumbnail(src, encoding):
    thumb_data, mime_type = src.getThumbnail(encoding=encoding)
    assert thumb_data
```

```
================================================= test session starts ==================================================
platform darwin -- Python 3.9.13, pytest-7.1.2, pluggy-1.0.0 -- /Users/bane/anaconda3/envs/dli-dev/bin/python
cachedir: .pytest_cache
rootdir: /Users/bane/Software/ResonantGeoData
plugins: anyio-3.6.1, factoryboy-2.5.0, mock-3.7.0, django-s3-file-field-0.3.0, Faker-13.13.0, django-4.5.2, cov-3.0.0
collected 21 items

test_thumbnail.py::test_thumbnail[JPEG] PASSED                                                                   [  4%]
test_thumbnail.py::test_thumbnail[PNG] PASSED                                                                    [  9%]
test_thumbnail.py::test_thumbnail[TIFF] PASSED                                                                   [ 14%]
test_thumbnail.py::test_thumbnail[TILED] PASSED                                                                  [ 19%]
test_thumbnail.py::test_thumbnail[JFIF] PASSED                                                                   [ 23%]
test_thumbnail.py::test_thumbnail[BMP] PASSED                                                                    [ 28%]
test_thumbnail.py::test_thumbnail[DIB] PASSED                                                                    [ 33%]
test_thumbnail.py::test_thumbnail[PCX] PASSED                                                                    [ 38%]
test_thumbnail.py::test_thumbnail[EPS] PASSED                                                                    [ 42%]
test_thumbnail.py::test_thumbnail[GIF] PASSED                                                                    [ 47%]
test_thumbnail.py::test_thumbnail[JPEG2000] PASSED                                                               [ 52%]
test_thumbnail.py::test_thumbnail[ICNS] PASSED                                                                   [ 57%]
test_thumbnail.py::test_thumbnail[ICO] PASSED                                                                    [ 61%]
test_thumbnail.py::test_thumbnail[MPO] PASSED                                                                    [ 66%]
test_thumbnail.py::test_thumbnail[PALM] FAILED                                                                   [ 71%]
test_thumbnail.py::test_thumbnail[PDF] PASSED                                                                    [ 76%]
test_thumbnail.py::test_thumbnail[PPM] PASSED                                                                    [ 80%]
test_thumbnail.py::test_thumbnail[SGI] PASSED                                                                    [ 85%]
test_thumbnail.py::test_thumbnail[TGA] PASSED                                                                    [ 90%]
test_thumbnail.py::test_thumbnail[WEBP] PASSED                                                                   [ 95%]
test_thumbnail.py::test_thumbnail[XBM] FAILED                                                                    [100%]

======================================================= FAILURES =======================================================
_________________________________________________ test_thumbnail[PALM] _________________________________________________

src = GDALFileTileSource ('/Users/bane/Desktop/rasters/converted.tif', 'JPEG', 95, 0, 'raw', False, '__STYLESTART__', None, '__STYLEEND__'),None,None,None,None
encoding = 'PALM'

    @pytest.mark.parametrize('encoding', TileOutputMimeTypes.keys())
    def test_thumbnail(src, encoding):
>       thumb_data, mime_type = src.getThumbnail(encoding=encoding)

test_thumbnail.py:13:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../kw/large_image/large_image/cache_util/cache.py:90: in wrapper
    v = func(self, *args, **kwargs)
../kw/large_image/sources/gdal/large_image_source_gdal/__init__.py:1022: in getThumbnail
    return super().getThumbnail(width, height, **kwargs)
../kw/large_image/large_image/cache_util/cache.py:90: in wrapper
    v = func(self, *args, **kwargs)
../kw/large_image/large_image/tilesource/base.py:1619: in getThumbnail
    return self.getRegion(**params)
../kw/large_image/sources/gdal/large_image_source_gdal/__init__.py:1181: in getRegion
    return super().getRegion(format, **kwargs)
../kw/large_image/large_image/tilesource/base.py:1789: in getRegion
    return _encodeImage(image, format=format, **kwargs)
../kw/large_image/large_image/tilesource/utilities.py:144: in _encodeImage
    imageData = _encodeImageBinary(
../kw/large_image/large_image/tilesource/utilities.py:102: in _encodeImageBinary
    image.save(output, encoding, **params)
../../anaconda3/envs/dli-dev/lib/python3.9/site-packages/PIL/Image.py:2300: in save
    save_handler(self, fp, filename)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

im = <PIL.Image.Image image mode=RGB size=256x256 at 0x111E61580>, fp = <_io.BytesIO object at 0x1121615e0>
filename = ''

    def _save(im, fp, filename):

        if im.mode == "P":

            # we assume this is a color Palm image with the standard colormap,
            # unless the "info" dict has a "custom-colormap" field

            rawmode = "P"
            bpp = 8
            version = 1

        elif im.mode == "L":
            if im.encoderinfo.get("bpp") in (1, 2, 4):
                # this is 8-bit grayscale, so we shift it to get the high-order bits,
                # and invert it because
                # Palm does greyscale from white (0) to black (1)
                bpp = im.encoderinfo["bpp"]
                im = im.point(
                    lambda x, shift=8 - bpp, maxval=(1 << bpp) - 1: maxval - (x >> shift)
                )
            elif im.info.get("bpp") in (1, 2, 4):
                # here we assume that even though the inherent mode is 8-bit grayscale,
                # only the lower bpp bits are significant.
                # We invert them to match the Palm.
                bpp = im.info["bpp"]
                im = im.point(lambda x, maxval=(1 << bpp) - 1: maxval - (x & maxval))
            else:
                raise OSError(f"cannot write mode {im.mode} as Palm")

            # we ignore the palette here
            im.mode = "P"
            rawmode = "P;" + str(bpp)
            version = 1

        elif im.mode == "1":

            # monochrome -- write it inverted, as is the Palm standard
            rawmode = "1;I"
            bpp = 1
            version = 0

        else:

>           raise OSError(f"cannot write mode {im.mode} as Palm")
E           OSError: cannot write mode RGB as Palm

../../anaconda3/envs/dli-dev/lib/python3.9/site-packages/PIL/PalmImagePlugin.py:157: OSError
_________________________________________________ test_thumbnail[XBM] __________________________________________________

src = GDALFileTileSource ('/Users/bane/Desktop/rasters/converted.tif', 'JPEG', 95, 0, 'raw', False, '__STYLESTART__', None, '__STYLEEND__'),None,None,None,None
encoding = 'XBM'

    @pytest.mark.parametrize('encoding', TileOutputMimeTypes.keys())
    def test_thumbnail(src, encoding):
>       thumb_data, mime_type = src.getThumbnail(encoding=encoding)

test_thumbnail.py:13:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../kw/large_image/large_image/cache_util/cache.py:90: in wrapper
    v = func(self, *args, **kwargs)
../kw/large_image/sources/gdal/large_image_source_gdal/__init__.py:1022: in getThumbnail
    return super().getThumbnail(width, height, **kwargs)
../kw/large_image/large_image/cache_util/cache.py:90: in wrapper
    v = func(self, *args, **kwargs)
../kw/large_image/large_image/tilesource/base.py:1619: in getThumbnail
    return self.getRegion(**params)
../kw/large_image/sources/gdal/large_image_source_gdal/__init__.py:1181: in getRegion
    return super().getRegion(format, **kwargs)
../kw/large_image/large_image/tilesource/base.py:1789: in getRegion
    return _encodeImage(image, format=format, **kwargs)
../kw/large_image/large_image/tilesource/utilities.py:144: in _encodeImage
    imageData = _encodeImageBinary(
../kw/large_image/large_image/tilesource/utilities.py:102: in _encodeImageBinary
    image.save(output, encoding, **params)
../../anaconda3/envs/dli-dev/lib/python3.9/site-packages/PIL/Image.py:2300: in save
    save_handler(self, fp, filename)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

im = <PIL.Image.Image image mode=RGB size=256x256 at 0x1123BAFA0>, fp = <_io.BytesIO object at 0x111b0b680>
filename = ''

    def _save(im, fp, filename):

        if im.mode != "1":
>           raise OSError(f"cannot write mode {im.mode} as XBM")
E           OSError: cannot write mode RGB as XBM

../../anaconda3/envs/dli-dev/lib/python3.9/site-packages/PIL/XbmImagePlugin.py:73: OSError
=============================================== short test summary info ================================================
FAILED test_thumbnail.py::test_thumbnail[PALM] - OSError: cannot write mode RGB as Palm
FAILED test_thumbnail.py::test_thumbnail[XBM] - OSError: cannot write mode RGB as XBM
============================================= 2 failed, 19 passed in 1.85s =============================================
```

`django-larege-image` is failing for these formats:

```
FAILED example/core/tests/test_data.py::test_thumbnail[pcx] - ValueError: Can...
FAILED example/core/tests/test_data.py::test_thumbnail[eps] - ValueError: ima...
FAILED example/core/tests/test_data.py::test_thumbnail[mpo] - OSError: cannot...
FAILED example/core/tests/test_data.py::test_thumbnail[palm] - OSError: canno...
FAILED example/core/tests/test_data.py::test_thumbnail[pdf] - ValueError: can...
FAILED example/core/tests/test_data.py::test_thumbnail[xbm] - OSError: cannot...
```

At the very least, I would expect opening the image with the same encoding to work. E.g.:

```py
src = large_image.open('/Users/bane/Desktop/rasters/converted.tif', encoding='PALM')
thumb_data, mime_type = src.getThumbnail(encoding='PALM')
```

```
_________________________________________________ test_thumbnail[PALM] _________________________________________________

encoding = 'PALM'

    @pytest.mark.parametrize('encoding', TileOutputMimeTypes.keys())
    def test_thumbnail(encoding):
        src = large_image.open('/Users/bane/Desktop/rasters/converted.tif', encoding=encoding)
>       thumb_data, mime_type = src.getThumbnail(encoding=encoding)

test_thumbnail.py:14:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../kw/large_image/large_image/cache_util/cache.py:90: in wrapper
    v = func(self, *args, **kwargs)
../kw/large_image/sources/gdal/large_image_source_gdal/__init__.py:1022: in getThumbnail
    return super().getThumbnail(width, height, **kwargs)
../kw/large_image/large_image/cache_util/cache.py:90: in wrapper
    v = func(self, *args, **kwargs)
../kw/large_image/large_image/tilesource/base.py:1619: in getThumbnail
    return self.getRegion(**params)
../kw/large_image/sources/gdal/large_image_source_gdal/__init__.py:1181: in getRegion
    return super().getRegion(format, **kwargs)
../kw/large_image/large_image/tilesource/base.py:1789: in getRegion
    return _encodeImage(image, format=format, **kwargs)
../kw/large_image/large_image/tilesource/utilities.py:144: in _encodeImage
    imageData = _encodeImageBinary(
../kw/large_image/large_image/tilesource/utilities.py:102: in _encodeImageBinary
    image.save(output, encoding, **params)
../../anaconda3/envs/dli-dev/lib/python3.9/site-packages/PIL/Image.py:2300: in save
    save_handler(self, fp, filename)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

im = <PIL.Image.Image image mode=RGB size=256x256 at 0x1971F0880>, fp = <_io.BytesIO object at 0x1974d8a90>
filename = ''

    def _save(im, fp, filename):

        if im.mode == "P":

            # we assume this is a color Palm image with the standard colormap,
            # unless the "info" dict has a "custom-colormap" field

            rawmode = "P"
            bpp = 8
            version = 1

        elif im.mode == "L":
            if im.encoderinfo.get("bpp") in (1, 2, 4):
                # this is 8-bit grayscale, so we shift it to get the high-order bits,
                # and invert it because
                # Palm does greyscale from white (0) to black (1)
                bpp = im.encoderinfo["bpp"]
                im = im.point(
                    lambda x, shift=8 - bpp, maxval=(1 << bpp) - 1: maxval - (x >> shift)
                )
            elif im.info.get("bpp") in (1, 2, 4):
                # here we assume that even though the inherent mode is 8-bit grayscale,
                # only the lower bpp bits are significant.
                # We invert them to match the Palm.
                bpp = im.info["bpp"]
                im = im.point(lambda x, maxval=(1 << bpp) - 1: maxval - (x & maxval))
            else:
                raise OSError(f"cannot write mode {im.mode} as Palm")

            # we ignore the palette here
            im.mode = "P"
            rawmode = "P;" + str(bpp)
            version = 1

        elif im.mode == "1":

            # monochrome -- write it inverted, as is the Palm standard
            rawmode = "1;I"
            bpp = 1
            version = 0

        else:

>           raise OSError(f"cannot write mode {im.mode} as Palm")
E           OSError: cannot write mode RGB as Palm

../../anaconda3/envs/dli-dev/lib/python3.9/site-packages/PIL/PalmImagePlugin.py:157: OSError
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -12,6 +12,7 @@
 - Better ignore tiff directories that aren't part of the pyramid ([943](../../pull/943))
 - Fix an issue with styling frames in ome tiffs ([945](../../pull/945))
 - Better handle large user records in annotation elements ([949](../../pull/949))
+- Harden which output formats are supported ([950](../../pull/950))
 
 ### Changes
 - Adjusted rest request logging rates for region endpoint ([948](../../pull/948))

--- a/large_image/tilesource/utilities.py
+++ b/large_image/tilesource/utilities.py
@@ -99,7 +99,19 @@ def _encodeImageBinary(image, encoding, jpegQuality, jpegSubsampling, tiffCompre
     elif encoding == 'PNG':
         params['compress_level'] = 2
     output = io.BytesIO()
-    image.save(output, encoding, **params)
+    try:
+        image.save(output, encoding, **params)
+    except Exception:
+        retry = True
+        if image.mode not in {'RGB', 'L'}:
+            image = image.convert('RGB')
+            try:
+                image.convert('RGB').save(output, encoding, **params)
+                retry = False
+            except Exception:
+                pass
+        if retry:
+            image.convert('1').save(output, encoding, **params)
     return ImageBytes(
         output.getvalue(),
         mimetype=f'image/{encoding.lower().replace("tiled", "tiff")}'
@@ -931,6 +943,10 @@ def addPILFormatsToOutputOptions():
     # Call this to actual register the extensions
     PIL.Image.registered_extensions()
     for key, value in PIL.Image.MIME.items():
+        # We don't support these formats; ICNS and ICO have fixed sizes; PALM
+        # and PDF can't be read back by PIL without extensions
+        if key in {'ICNS', 'ICO', 'PALM', 'PDF'}:
+            continue
         if key not in TileOutputMimeTypes and key in PIL.Image.SAVE:
             TileOutputMimeTypes[key] = value
     for key, value in PIL.Image.registered_extensions().items():

--- a/test/test_source_base.py
+++ b/test/test_source_base.py
@@ -544,22 +544,22 @@ def testImageBytes():
     assert ib._repr_png_() is None
 
 
-def testOutputFormats():
+@pytest.mark.parametrize('format', [
+    format for format in large_image.constants.TileOutputMimeTypes
+    if format not in {'TILED'}])
+def testOutputFormats(format):
     imagePath = datastore.fetch('sample_image.ptif')
     testDir = os.path.dirname(os.path.realpath(__file__))
     imagePathRGBA = os.path.join(testDir, 'test_files', 'rgba_geotiff.tiff')
-    for format in large_image.constants.TileOutputMimeTypes:
-        if format == 'TILED':
-            continue
 
-        ts = large_image.open(imagePath, encoding=format)
-        img = PIL.Image.open(io.BytesIO(ts.getTile(0, 0, 0)))
-        assert (img.width, img.height) == (256, 256)
-        img = PIL.Image.open(io.BytesIO(ts.getThumbnail(encoding=format)[0]))
-        assert (img.width, img.height) == (256, 53)
+    ts = large_image.open(imagePath, encoding=format)
+    img = PIL.Image.open(io.BytesIO(ts.getTile(0, 0, 0)))
+    assert (img.width, img.height) == (256, 256)
+    img = PIL.Image.open(io.BytesIO(ts.getThumbnail(encoding=format)[0]))
+    assert (img.width, img.height) == (256, 53)
 
-        ts = large_image.open(imagePathRGBA, encoding=format)
-        img = PIL.Image.open(io.BytesIO(ts.getTile(0, 0, 0)))
-        assert (img.width, img.height) == (256, 256)
-        img = PIL.Image.open(io.BytesIO(ts.getThumbnail(encoding=format)[0]))
-        assert (img.width, img.height) == (256, 256)
+    ts = large_image.open(imagePathRGBA, encoding=format)
+    img = PIL.Image.open(io.BytesIO(ts.getTile(0, 0, 0)))
+    assert (img.width, img.height) == (256, 256)
+    img = PIL.Image.open(io.BytesIO(ts.getThumbnail(encoding=format)[0]))
+    assert (img.width, img.height) == (256, 256)

--- a/test/test_source_base.py
+++ b/test/test_source_base.py
@@ -1,8 +1,10 @@
+import io
 import os
 import re
 import sys
 from pathlib import Path
 
+import PIL.Image
 import pytest
 
 import large_image
@@ -540,3 +542,24 @@ def testImageBytes():
     assert 'ImageBytes' in repr(ib)
     assert ib._repr_jpeg_() is None
     assert ib._repr_png_() is None
+
+
+def testOutputFormats():
+    imagePath = datastore.fetch('sample_image.ptif')
+    testDir = os.path.dirname(os.path.realpath(__file__))
+    imagePathRGBA = os.path.join(testDir, 'test_files', 'rgba_geotiff.tiff')
+    for format in large_image.constants.TileOutputMimeTypes:
+        if format == 'TILED':
+            continue
+
+        ts = large_image.open(imagePath, encoding=format)
+        img = PIL.Image.open(io.BytesIO(ts.getTile(0, 0, 0)))
+        assert (img.width, img.height) == (256, 256)
+        img = PIL.Image.open(io.BytesIO(ts.getThumbnail(encoding=format)[0]))
+        assert (img.width, img.height) == (256, 53)
+
+        ts = large_image.open(imagePathRGBA, encoding=format)
+        img = PIL.Image.open(io.BytesIO(ts.getTile(0, 0, 0)))
+        assert (img.width, img.height) == (256, 256)
+        img = PIL.Image.open(io.BytesIO(ts.getThumbnail(encoding=format)[0]))
+        assert (img.width, img.height) == (256, 256)

